File: python/reverse-odd-levels-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def reverse_level(q: List[Optional[TreeNode]], n: int):
        for i in range(n // 2):
            j = n - i - 1
            q[i].val, q[j].val = q[j].val, q[i].val

    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        q = [root]
        level = 0
        while q:
            n = len(q)
            if level & 1 == 1:
                Solution.reverse_level(q, n)
            for _ in range(n):
                node = q.pop(0)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            level += 1
        return root

File: go/reverse-odd-levels-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func reverseLevel(q []*TreeNode, n int) []*TreeNode {
	for i := 0; i < (n / 2); i++ {
		j := n - i - 1
		aux := q[i].Val
		q[i].Val = q[j].Val
		q[j].Val = aux
	}
	return q
}

func addNode(q []*TreeNode, node *TreeNode) []*TreeNode {
	if node != nil {
		q = append(q, node)
	}
	return q
}

func reverseOddLevels(root *TreeNode) *TreeNode {
	q := make([]*TreeNode, 0)
	q = append(q, root)
	level := 0
	for 0 < len(q) {
		n := len(q)
		if level&1 == 1 {
			q = reverseLevel(q, n)
		}
		for i := 0; i < n; i++ {
			node := q[0]
			q = q[1:]
			q = addNode(q, node.Left)
			q = addNode(q, node.Right)
		}
		level += 1
	}
	return root
}
